src/ess/powder/correction.py

```diff
@@ -257,6 +257,86 @@ def _shallow_copy(da: sc.DataArray) -> sc.DataArray:
     return out
 
 
+def compute_pdf_from_structure_factor(
+    s: sc.DataArray,
+    *,
+    rho: sc.Variable,
+    r: sc.Variable,
+    self_scattering: sc.Variable,
+    use_filter: bool,
+) -> sc.DataArray:
+    '''
+    Computes the pair-distribution-function :math:`D(r)` as defined in
+    `Review: Pair distribution functions from neutron total scattering for the study of local structure in disordered materials <https://www.sciencedirect.com/science/article/pii/S2773183922000374>`_
+    from the overall scattering function :math:`S(Q)`.
+
+    The inputs to the algorithm are:
+
+    * A histogram representing :math:`S(Q)` with :math:`N` bins on a bin-edge grid with :math:`N+1` edges :math:`Q_j` for :math:`j=0\\ldots N`.
+    * The density :math:`\\rho`, the total number of atoms per volume.
+    * The structure independent coherent scattering amplitude :math:`S_{self} = \\sum_m c_m b_m^2` where the sum is over the atomic species in the sample, :math:`c_m` is the fraction of atoms that belong to species :math:`m` and :math:`b_m` is the size of scattering from an atom of that species.
+    * The bin-edge grid over :math:`r` the output histogram representing :math:`D(r)` will be computed on.
+
+    In each output bin, the output is computed as:
+
+    .. math::
+        D_{i+\\frac{1}{2}} = \\frac{2}{\\pi(r_{i+1}-r_i)} \\int_{r_i}^{r_{i+1}} \\int_{0}^\\infty (S(Q) - S_{self}) Q sin(Q r) dQ \\ dr  \\\\
+        \\approx \\frac{2}{\\pi(r_{i+1}-r_i)} \\sum_{j=0}^{N-1} (S(Q)_{j+\\frac{1}{2}} - S_{self}) (cos(\\bar{Q}_{j+\\frac{1}{2}} r_{i})-cos(\\bar{Q}_{j+\\frac{1}{2}} r_{i+1})) \\Delta Q_{j+\\frac{1}{2}}
+
+    Note that in the above expression the subscript :math:`_{j+\\frac{1}{2}}` is used to denote
+    quantities belonging to the :math:`j`:th bin of a histogram, :math:`\\bar{Q}_{j+\\frac{1}{2}} = \\frac{Q_j + Q_{j+1}}{2}` and :math:`\\Delta Q_{j+\\frac{1}{2}} = Q_{j+1} - Q_{j}`.
+
+    Parameters
+    ----------
+    s:
+        :math:`S(Q)` with bin-edge coordinate :math:`Q`
+    rho:
+        total number of atoms per unit volume
+    r:
+        bin-edges of output grid
+    self_scattering:
+        the amplitude of the overall scattering that is due to self-scattering
+    use_filter:
+        if ``True`` the Lorch filter is applied
+
+    Returns
+    -------
+    :
+        :math:`D(r)` for each bin in the provided output grid
+
+    '''  # noqa: E501
+    for i in range(s.size):
+        if not sc.isnan(s[i]).value:
+            minbound = i
+            break
+    for i in range(s.size - 1, -1, -1):
+        if not sc.isnan(s[i]).value:
+            maxbound = i
+            break
+
+    s = s[minbound : maxbound + 1]
+    q = s.coords['Q']
+    qm = sc.midpoints(q)
```

Comment:
This is only correct for linearly spaced bins. Handling other types of bins is difficult here because you need both bin centres and bin edges. So please add a `sc.islinspace(q)` check and raise an exception. We can implement support for other bins later if needed.

Reply:
What do you mean by "correct"?

I don't think an exception is the best way to go here. Using the midpoint is not incorrect for non-equidistant q if by correct we mean "as the q-binning is refined, the result converges to the exact integral".
Using the midpoint here is an approximation both when the q edges are linspaced and when they are unequally spaced. It's a *better* approximation (in a sense) when the q edges are linspaced.

Requiring linspaced grid here limits the usability of the routine without any quantifiable correctness improvement.

Reply:
Ok. Leave it